### ISSUE TITLE
binance js undefined error

### ISSUE THREAD

**Issue by dwarda** [askmike/gekko#1712]
**Note: for support questions, please join our [Discord server](https://discord.gg/26wMygt)**

* **I'm submitting a ...**
[x] bug report
[ ] feature request
[ ] question about the decisions made in the repository

* **Action taken** (I tried to delete and reinstall again)

* **Expected result** (hoping it would trade on binance without giving me an error)

* **Actual result** (same error keeps poping up)

Hi everyone,
I am sorry to bring up this error again. from reading your comment it feels like it has been solved but I know nothing about programming and have no idea what you guys are talking about.
every time I run Gekko on and try to trade on Binance,  I get this error after less than ten min:

RECEIVED ERROR IN GEKKO 0707182042988577

Gekko encountered an error and can't continue

Gekko encountered an error and can't continue

Error:

[binance.js] undefined

Meta debug info:

Gekko version: v0.5.12
Nodejs version: v8.9.4

I would really appreciate it if you explain to me in step what to do in order to fix it.
thank you in advance

**Comment by cmroche** [comment]
Try `npm uninstall binance` then `npm install binance` so the error print properly, then let me know what the error is please :)

**Comment by ghost** [comment]
@cmroche - similar issue - with 1 minute candle sizes on binance when trading XMR there seems to be issues with the orders not filling and it cancels/retries a lot. Any chance you could fix this? I'm using your product_update branch. If you leave me an eth address or email falatime@gmail.com I'll send a little something. Much appreciated!

**Comment by dwarda** [comment]
@cmroche - here is the error that im receiving after doing what you said: 

RECEIVED ERROR IN GEKKO 7869689313950996


Gekko encountered an error and can't continue


Gekko encountered an error and can't continue

Error:

[binance.js] Response code 400

Meta debug info:

Gekko version: v0.5.12
Nodejs version: v8.9.4

I am getting this error after about 10 min from starting trading BTC and TRX on binance using RSI with these parameters:
interval = 16

[thresholds]
low = 21
high = 70
persistence = 1
candle size (3 min) and warm up time (10) 

thank you

**Comment by cmroche** [comment]
Error code 400 means there is a data error. @mvanbosc @dwarda are you using pull request #1550 for this? If not then I suspect you might have bad market configurations and are not sending the correct lot sizes in the buy and sell orders. Please try that PR and let me know if it corrects your issues.

Also, can you turn on debug in your config files to see what is being passed to the exchange. If you are using the UI modify this option in web/routes/baseConfig.js, set config.debug to true

**Comment by cmroche** [comment]
Oh sorry i just realized what this is, it’s your time synch.. people using the binance API are starting to report this problems in large number, look at my answer at issue #1663 for the solution. Your internal clock has drifted by more than 1 second now you need to set it up to sync more than once a day.

TLDR do what it says here

http://www.thewindowsclub.com/change-internet-time-update-interval-windows

https://www.opentechguides.com/how-to/article/windows-10/121/windows-10-internet-time.html

**Comment by dwarda** [comment]
@cmroche Thank you for your support man. but your links are for windows 10. does it matter that im using a mac ?

**Comment by cmroche** [comment]
A little bit, but OS X does have settings for this so where, worth checking. Sorry I cannot tell you offhand where time should synchronization is setup. 

If you do not think this is the issue in your case enable debug logging and send the full output when you get the error please.

**Comment by dwarda** [comment]
@cmroche I tried all of the suggestion in the other link. uninstalling and re-installing binance. changing some numbers in binance.js as suggested but nothing seems to work. based on time.gov my mac time doesn't seem to be drifting. I am still getting the same error. so as you said, I don't think the time drifting is the issue I'm having. however I don't really know anything about programming, so I would really appreciate it if you tell me how to enable debug logging and where can I find the file that I need to send you. 

thank you so much

**Comment by dwarda** [comment]
@cmroche I think I found the error that you asked for. 
I hope this is it 
[2018-01-18-19-45-UTC-tradebot.log](https://github.com/askmike/gekko/files/1644419/2018-01-18-19-45-UTC-tradebot.log)

**Comment by storshun** [comment]
@dwarda go to this website and see what it says your time drift is:
https://time.is/
It's located near the top left. If the number is 0.300 or larger, your time is too far off and you'll have these errors. Use a program like Dimension4 to select a time server, click the "settings" button and choose "How Often" and set it to "Every" and 1 minute.
![image](https://user-images.githubusercontent.com/35028707/35135351-5398ac20-fca2-11e7-8302-a33578e1017b.png)
![image](https://user-images.githubusercontent.com/35028707/35135373-746de0a0-fca2-11e7-915f-2e80e6df2cb8.png)

Make sure the server you choose gives you a synchronized response like mine does in the prior image.

This solved 100% of my problems that were like yours once I was on the latest install of npm binance, and was using a recent version of gekko.

**Comment by dwarda** [comment]
@srotshun, thank you for your input. Based on that website, my Mac is 0.18 behind in time, is that significant ? And Dimension 4 is only for windows, it didn’t work on my MacBook. I wasn’t able to find a sync app for Mac. Do you know any ?

**Comment by christhecanadien** [comment]
Hi Guys, I'm having the same issue: [binance.js] Response code 400
According with Time.is my Mac is -0,006 sec drifted !!! Is it a big problem ?
Please heeeeelp !

**Comment by cmroche** [comment]
There is a change now pushed to npm to enable Binance server synchronization. Part of the problem may simply be that Binance's time is also a little bit out exacerbating the issue. I'll integrate these changes and add the new feature to our Binance implementation this afternoon and provide the PR once ready.

See here: https://github.com/aarongarvey/binance/issues/23

**Comment by bakman2** [comment]
FYI if you accessed the binance api before with another tool, the nonce time could be off by a factor 1000 (microseconds vs milliseconds) - this requires you to generate new keys. 
For ntp use pool.ntp.org

**Comment by askmike** [comment]
@dwarda: @cmroche's fix has been merged into develop, please try that and report back if the problem still persists.
